### PR TITLE
For next

### PR DESCRIPTION
**Pull request by hclee**

### CODE REVIEW

File: fsck/fsck.c [hunk 1]
```diff
@@ -670,58 +670,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
...
+				  bd->num_sectors);
 		goto err;
 	}
 
 	if (le32_to_cpu(bs->bsx.clu_count) * EXFAT_CLUSTER_SIZE(bs) >
 			bd->size) {
-		exfat_err("too large cluster count: %u, expected: %u\n",
-				le32_to_cpu(bs->bsx.clu_count),
-				bd->num_clusters);
+		if (main_sect)
```
Comment (namjaejeon) [inline fsck/fsck.c:735]: why does it print error message only if main_sect is true ?

Comment (hclee) [reply fsck/fsck.c:735]: Because this function is called many times to read and verify the backup boot region with different sector sizes for restoring the main boot region, these error messages is too verbose for this restoration.

File: fsck/fsck.c [hunk 2]
```diff
@@ -670,58 +670,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
...
 
 	if (memcmp(bs->bpb.oem_name, "EXFAT   ", 8) != 0) {
-		exfat_err("failed to find exfat file system.\n");
+		if (main_sect)
+			exfat_err("failed to find exfat file system\n");
 		goto err;
 	}
 
-	ret = boot_region_checksum(bd, bs_offset);
+	ret = boot_region_checksum(bd, bs_offset, sect_size);
```
Comment (namjaejeon) [inline fsck/fsck.c:691]: If bd->sector_size is no long used, we don't need to make struct bd as argument of boot_region_checksum(). just give bd->dev_fd as an agument ?

File: fsck/fsck.c (hunk 1 above)
Comment (namjaejeon) [reply fsck/fsck.c:735]: I think that we should add bool silent variable as an argument of this function.

File: dump/dump.c [hunk 3]
```diff
@@ -64,26 +64,34 @@ static unsigned int exfat_count_used_clusters(unsigned char *bitmap,
 	return count;
 }
 
+
+static off_t c2o(off_t heap_offset, unsigned int clu, unsigned int clu_size)
```
Comment (namjaejeon) [inline dump/dump.c:68]: we don't need to prefix(exfat_) for this function ?

File: dump/dump.c [hunk 4]
```diff
@@ -64,26 +64,34 @@ static unsigned int exfat_count_used_clusters(unsigned char *bitmap,
...
-	unsigned int total_clus, used_clus, clu_offset, root_clu;
+	unsigned int total_clus, used_clus, root_clu;
+	off_t heap_offset;
 	unsigned long long bitmap_len;
 	int ret;
 	unsigned char *bitmap;
 	char *volume_label;
 
-	ppbr = malloc(bd->sector_size);
+	ppbr = malloc(sizeof(*ppbr));
```
Comment (namjaejeon) [inline dump/dump.c:87]: As I know, we should access  sector as per 4k unit in case of native 4K sector.(i.e. hw sector size : 4K, logical sector size : 4K)

File: fsck/fsck.c [hunk 5]
```diff
@@ -670,58 +670,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
...
 
 	ret = -EINVAL;
 	if (EXFAT_SECTOR_SIZE(bs) < 512 || EXFAT_SECTOR_SIZE(bs) > 4 * KB) {
-		exfat_err("too small or big sector size: %d\n",
-				EXFAT_SECTOR_SIZE(bs));
+		if (main_sect)
+			exfat_err("too small or big sector size: %d\n",
+				  EXFAT_SECTOR_SIZE(bs));
 		goto err;
 	}
```
Comment (namjaejeon) [inline fsck/fsck.c:702]: I agreed Eric's suggestion that update bd->sector_size here. I think that's the best way to do it.

File: dump/dump.c (hunk 4 above)
Comment (namjaejeon) [reply dump/dump.c:87]: I will declare EXFAT_MAX_SECTOR_SIZE 4096 macro, we can use it here.

File: fsck/fsck.c (hunk 1 above)
Comment (hclee) [reply fsck/fsck.c:735]: I will declare a local boolean variable for that.

File: fsck/fsck.c (hunk 2 above)
Comment (hclee) [reply fsck/fsck.c:691]: Replacing bd with bd->dev_fd seems good. But I don't think the bd parameter is strange, and the sect_size parameter means Exfat's sector size.

File: dump/dump.c (hunk 3 above)
Comment (hclee) [reply dump/dump.c:68]: This function is static, So I don't append the prefix "exfat_".

File: dump/dump.c (hunk 4 above)
Comment (hclee) [reply dump/dump.c:87]: Okay, I will change this.

File: fsck/fsck.c (hunk 5 above)
Comment (hclee) [reply fsck/fsck.c:702]: I think  Eric's suggestion can cause the confusion about device's sector size and Exfat's sector size, But it is the easiest way to solve the problem which comes up when device's sector size and Exfat's sector size are different. So I think we can apply the suggestion right now. But fsck does not need that, because fsck does not depend on the device sector size.

File: dump/dump.c (hunk 4 above)
Comment (hclee) [reply dump/dump.c:87]: Ah, Reading small data with read(2) can be a problem? read(2) is buffered I/O. Let me check this.

File: fsck/fsck.c (hunk 1 above)
Comment (namjaejeon) [reply fsck/fsck.c:735]: I don't know how you do that.

File: fsck/fsck.c (hunk 2 above)
Comment (namjaejeon) [reply fsck/fsck.c:691]: I have nothing to say.

File: dump/dump.c (hunk 3 above)
Comment (namjaejeon) [reply dump/dump.c:68]: Do whatever you want. but all function has exfat prefix.

File: dump/dump.c (hunk 4 above)
Comment (namjaejeon) [reply dump/dump.c:87]: Do you have 4K native device ? you should read under 4KB against block device, not filesystem.

File: fsck/fsck.c (hunk 5 above)
Comment (namjaejeon) [reply fsck/fsck.c:702]: if so, we should not use struct bd->dev_fd also. Do you think it is not confused ? we should think exfat image as bd.

Comment (namjaejeon) [reply fsck/fsck.c:702]: > because fsck does not depend on the device sector size.
If you used struct bd in fsck, please don't make more complex change, which mean we don't need to separate bd->dev_fd and exfat sector size.

File: dump/dump.c (hunk 4 above)
Comment (hclee) [reply dump/dump.c:87]: We read data from a block device file which is one of file in "bdev" filesystem. Anyway I will apply this and the above comments

File: fsck/fsck.c (hunk 5 above)
Comment (hclee) [reply fsck/fsck.c:702]: Okay, this is simple.

**Comment by hclee** [comment]
@namjaejeon I sent a new PR. please check this.

File: fsck/fsck.c [hunk 6]
```diff
@@ -670,58 +671,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
...
-			bs_offset * bd->sector_size) != (ssize_t)sizeof(*bs)) {
+			bs_offset * sect_size) != (ssize_t)sizeof(*bs)) {
 		exfat_err("failed to read a boot sector\n");
 		ret = -EIO;
 		goto err;
 	}
 
 	if (memcmp(bs->bpb.oem_name, "EXFAT   ", 8) != 0) {
-		exfat_err("failed to find exfat file system.\n");
+		if (main_sect)
```
Comment (namjaejeon) [inline fsck/fsck.c:687]: Why do you not use bool silent here?

File: fsck/fsck.c [hunk 7]
```diff
@@ -670,58 +671,72 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
 		return -ENOMEM;
 	}
 
+	if (main_sect)
```
Comment (namjaejeon) [inline fsck/fsck.c:674]: main_sect is needed ? you can use bs_offset as argument like before

File: fsck/fsck.c [hunk 8]
```diff
@@ -775,24 +790,51 @@ static int restore_boot_region(struct exfat_blk_dev *bd)
...
-			exfat_err("backup boot region is also corrupted\n");
-			return ret;
-		}
-		ret = restore_boot_region(exfat->blk_dev);
-		if (ret < 0) {
-			exfat_err("failed to restore boot region from backup\n");
-			free(*bs);
-			*bs = NULL;
-			return ret;
+		const unsigned int sector_sizes[] = {512, 4096, 1024, 2048};
```
Comment (namjaejeon) [inline fsck/fsck.c:816]: we have got exfat sector size from boot sector above. Why do we need to check it using 4 kinds of sector size ?

File: fsck/fsck.c (hunk 6 above)
Comment (hclee) [reply fsck/fsck.c:687]: it is a mistake.

File: fsck/fsck.c (hunk 7 above)
Comment (hclee) [reply fsck/fsck.c:674]: Okay, I will change it.

File: fsck/fsck.c (hunk 8 above)
Comment (hclee) [reply fsck/fsck.c:816]: Because we don't trust the sector size of the main boot sector. it seems better to start the loop with the sector size of the main if it is valid.

File: fsck/fsck.c [hunk 9]
```diff
@@ -732,31 +741,31 @@ static int read_boot_region(struct exfat_blk_dev *bd, struct pbr **pbr,
...
+	sector = malloc(sect_size);
 	if (!sector)
 		return -ENOMEM;
 
 	for (i = 0; i < 12; i++) {
-		if (exfat_read(bd->dev_fd, sector, bd->sector_size,
-				BACKUP_BOOT_SEC_IDX * bd->sector_size +
-				i * bd->sector_size) !=
-				(ssize_t)bd->sector_size) {
+		if (exfat_read(bd->dev_fd, sector, sect_size,
```
Comment (namjaejeon) [inline fsck/fsck.c:755]: I still think this is weird. If you update bd->sector_size on caller, There is no need to make unnecessary changes.

Comment (hclee) [reply fsck/fsck.c:755]: this function could be called many times to read the backup boot region. I think sect_size parameter is better than updating bd->sector_size for each call. And in fsck, bd->sector_size is not referenced anymore except this function.

File: fsck/fsck.c (hunk 8 above)
Comment (namjaejeon) [reply fsck/fsck.c:811]: main boot sector should be recover using backup's one? You are saying backup boot sector can even not be trusted ? Have you checked a such case with windows fsck ? I want to know whether windows fsck can recover or not.

**Comment by hclee** [comment]
I am implementing repairs, And I will check if the main boot sector can be
recovered without the backup boot sector.

2021년 5월 17일 (월) 오전 9:15, Namjae Jeon ***@***.***>님이 작성:
